src/telemetry/session-stats/SessionActivityController.ts

```diff
@@ -0,0 +1,402 @@
+/**
+ * External dependencies
+ */
+import { select } from '@wordpress/data';
+import * as Y from 'yjs';
+
+/**
+ * Internal dependencies
+ */
+import { SessionStats } from './SessionStats';
+import { SessionStatsTelemetryLogger } from './SessionStatsTelemetryLogger';
+import { UserInactivityTimer } from './UserInactivityTimer';
+import { SESSION_STATS_ORIGIN, SESSION_TIMEOUT_IN_SEC } from './constants';
+import {
+	areUserSetsEqual,
+	getConnectedUserCount,
+	getConnectedUserIds,
+	isPositiveInteger,
+} from './utils';
+import { store as awarenessStore } from '@/store/awareness-store';
+import { Logger } from '@/utilities/logger';
+
+import type { Awareness } from 'y-protocols/awareness';
+
+/**
+ * Coordinates collaborative session events and the execution of operations in
+ * response to them.
+ *
+ * Subscribes to session events, manages session state transitions, tracks user
+ * activity, responds to user and document changes, handles inactivity, logs
+ * session data.
+ */
+export class SessionActivityController {
+	private static __instance?: SessionActivityController;
+
+	private inactivityTimer: UserInactivityTimer;
+	private logger = new Logger( SESSION_STATS_ORIGIN );
+	private sessionStats: SessionStats;
+	private isLoggingInProgress = false;
+	private logSessionTimeout?: NodeJS.Timeout;
+
+	private awarenessChangeHandler?: () => void;
+	private documentUpdateHandler?: (
+		update: Uint8Array,
+		origin: unknown,
+		doc: Y.Doc,
+		transaction: Y.Transaction
+	) => void;
+
+	/**
+	 * Creates a SessionActivityController instance, initializes the session
+	 * stats tracker and inactivity timer, and subscribes to user and document
+	 * change events.
+	 *
+	 * @param awareness The Yjs Awareness instance to use
+	 */
+	constructor( private awareness: Awareness ) {
+		this.sessionStats = new SessionStats( awareness, this.logger );
+
+		this.inactivityTimer = new UserInactivityTimer(
+			SESSION_TIMEOUT_IN_SEC * 1000,
+			() => this.sessionStats.getLastActivityTime(),
+			() => this.handleInactivityTimeout()
+		);
+
+		this.subscribeToUserChanges();
```

Comment:
Looking at this, I wonder if the class could be re-done such that:

- Instantiate it within the constructor of the awareness manager, so their lifecycles match. This means a single clientID at a time would have this class instantiated.
- Within the existing subscriptions, call in the methods on this class that require those subscriptions. That way, there's one single grouped awareness subscription for the awareness changes to be transmitted to. That was a performance optimization that had been mentioned on some forum posts (don't have the link on hand but I remember running into that as an improvement as more clients join).